src/Features/Core/Portable/CodeLens/CodeLensFindReferenceProgress.cs

```diff
@@ -0,0 +1,173 @@
+
+        public int ReferencesCount => _locations.Count;
+
+        public ImmutableArray<Location> Locations => _locations.ToImmutableArray();
+
+        public CodeLensFindReferencesProgress(
+            ISymbol queriedDefinition,
+            SyntaxNode queriedNode,
+            int searchCap,
+            CancellationToken cancellationToken)
+        {
+            _queriedSymbol = queriedDefinition;
+            _queriedNode = queriedNode;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _aggregateCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(
+                _cancellationTokenSource.Token, cancellationToken);
+            _locations = new ConcurrentSet<Location>();
+
+            SearchCap = searchCap;
+        }
+
+        public void OnCompleted()
+        {
+        }
+
+        /// <summary>
+        /// Returns partial symbol locations whose node does not match the given syntaxNode
+        /// </summary>
+        /// <param name="symbol">Symbol whose locations are queried</param>
+        /// <param name="syntaxNode">Syntax node to compare against to exclude location - actual location being queried</param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        /// <returns>Partial locations</returns>
+        private static IEnumerable<Location> GetPartialLocations(ISymbol symbol, SyntaxNode syntaxNode, CancellationToken cancellationToken)
+        {
+            // Returns nodes from source not equal to actual location and matches the definition locations to syntax references, ignores metadata locations
+            return from syntaxReference in symbol.DeclaringSyntaxReferences
+                   let candidateSyntaxNode = syntaxReference.GetSyntax(cancellationToken)
+                   where !(syntaxNode.Span == candidateSyntaxNode.Span &&
+                           syntaxNode.SyntaxTree.FilePath.Equals(candidateSyntaxNode.SyntaxTree.FilePath, StringComparison.OrdinalIgnoreCase))
+                   from sourceLocation in symbol.Locations
+                   where !sourceLocation.IsInMetadata
+                         && candidateSyntaxNode.SyntaxTree.Equals(sourceLocation.SourceTree)
+                         && candidateSyntaxNode.Span.Contains(sourceLocation.SourceSpan)
+                   select sourceLocation;
+        }
+
+        /// <summary>
+        /// Exclude the following kind of symbols:
+        ///  1. Implicitly declared symbols (such as implicit fields backing properties)
+        ///  2. Symbols that can't be referenced by name (such as property getters and setters).
+        ///  3. Metadata only symbols, i.e. symbols with no location in source.
+        /// </summary>
+        private static bool FilterReference(ISymbol queriedSymbol, ISymbol definition, ReferenceLocation reference)
+        {
+            return definition.IsImplicitlyDeclared ||
```

Comment:
on a personal level, i really dislike hugely complex expressions like this.  They're very annoying to debug through when there's a problem.  I'd far prefer a single if satement per clause with a nice clear explanation why we're filtering out those entities.  It also makes it much easier to augment in the future.